### PR TITLE
PP-1099: Add error checking in Scheduler for resources

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Issue-ID

* **[PP-1099](https://pbspro.atlassian.net/browse/PP-1099)**

#### Problem description, Cause and Analysis
I noticed that in some very rare instances, accounting E records end up having -2kb as the mem value in the exec_vnode field. While I've not been able to reproduce it, I'm pretty sure about why it's happening: Scheduler's set_resources_req() function calls res_to_num() to convert resource values into numeric. res_to_num() returns a numeric value for all resource types, except string, for string types, it returns SCHD_INFINITY, which translates to -2. So, if for some reason res_to_num() is not able to convert the mem value to a number, it will return SCHD_INFINITY. But because 'mem' is a consumable type, the scheduler looks at its numeric value while evaluating the job for placement, which means it'll think that the job is requesting -2kb mem, and end up creating the exec_vnode with it. So, set_resource_req() should check to see that a 'size' type attribute does not get set to SCHD_INFINITY.

#### Solution description
- Added a check inside set_resource_req() to verify that 'res_to_num()' returns SCHD_INFINITY for only string type resources.
- set_resource_req() now returns an int, 1 for Success, 0 for Error
- Some places in code now check for return value of set_resource_req() and error out if it returned an error.

#### Checklist:

- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/scheduler/job_info.c [hunk 1]
```diff
@@ -1109,29 +1109,33 @@ query_job(struct batch_status *job, server_info *sinfo, schd_error *err)
...
-			set_resource_req(resreq, attrp->value);
-
-			if (resresv->resreq == NULL)
+			if (set_resource_req(resreq, attrp->value) != 1) {
+				set_schd_error_codes(err, NEVER_RUN, ERR_SPECIAL);
+				set_schd_error_arg(err, SPECMSG, "bad resource requested data");
+				resresv->is_invalid = 1;
+			} else {
+				if (resresv->resreq == NULL)
 				resresv->resreq = resreq;
```
Comment (mike0042) [inline src/scheduler/job_info.c]: Line needs to be indented.

**Comment by bhroam** [comment]
@agrawalravi90 Travis failed on one test.  Please run the smoke test locally and debug this.

File: src/scheduler/misc.c [hunk 2]
```diff
@@ -250,11 +250,12 @@ add_str_to_array(char ***str_arr, char *str)
  * @param[in]	res_str	-	the resource string
  * @param[out]	type	-	the type of the resource
  *
- * @return	a number in kilobytes or seconds
+ * @return	sch_resource_t
+ * @retval	a number in kilobytes or seconds
```
Comment (bhroam) [inline src/scheduler/misc.c]: Can you change the @ brief to say something similar to this?

File: src/scheduler/misc.c [hunk 3]
```diff
@@ -250,11 +250,12 @@ add_str_to_array(char ***str_arr, char *str)
...
+ * @retval	0 for False, 1 for True
```
Comment (bhroam) [inline src/scheduler/misc.c]: Break this into two @ retval lines.

File: src/scheduler/misc.c [hunk 4]
```diff
@@ -906,12 +907,14 @@ is_num(char *str)
 		return 1;
 
 	/* is the string a size */
-	if (i == strlen(str) - 2) {
+	if ((i == strlen(str) - 2) || (i == strlen(str) - 1)) {
```
Comment (bhroam) [inline src/scheduler/misc.c]: Call strlen() once and save the value.

File: src/scheduler/misc.c [hunk 5]
```diff
@@ -906,12 +907,14 @@ is_num(char *str)
...
 		c = tolower(str[i]);
 		if (c == 'k' || c == 'm' || c == 'g' || c == 't') {
 			c = tolower(str[i+1]);
 			if (c == 'b' || c == 'w')
 				return 1;
+			else if (c == '\0')
```
Comment (bhroam) [inline src/scheduler/misc.c]: Combine this with the above if statement.

File: src/scheduler/resource_resv.c [hunk 6]
```diff
@@ -859,8 +859,13 @@ create_resource_req(char *name, char *value)
 			resreq->name = rdef->name;
 			resreq->type = rdef->type;
 
-			if (value != NULL)
-				set_resource_req(resreq, value);
+			if (value != NULL) {
+				if (set_resource_req(resreq, value) != 1) {
+					schdlog(PBSEVENT_DEBUG, PBS_EVENTCLASS_SCHED, LOG_DEBUG, name,
+						"Bad resource data found");
```
Comment (bhroam) [inline src/scheduler/resource_resv.c]: This log message is different than the above message.  I'd be consistent.

File: src/scheduler/resource_resv.c [hunk 7]
```diff
@@ -1004,25 +1009,35 @@ find_resource_req(resource_req *reqlist, resdef *def)
...
-	rdef = find_resdef(allres, req->name);
+		rdef = find_resdef(allres, req->name);
 		req->def = rdef;
 	}
 	if (rdef != NULL)
 		req->type = rdef->type;
+
+	if (req->amount == SCHD_INFINITY) {
+		/* Verify that this is actually a non-numeric resource */
+		if (!(req->def->type.is_string))
```
Comment (bhroam) [inline src/scheduler/resource_resv.c]: You don't necessarily need the inner set of ().  The -> and . operators have higher precedence than !

**Comment by agrawalravi90** [comment]
Thanks for the comments guys, I've addressed them, please let me know if I should change anything else.

File: src/scheduler/misc.c [hunk 8]
```diff
@@ -242,17 +242,23 @@ add_str_to_array(char ***str_arr, char *str)
 
 /**
  * @brief
- * 		res_to_num - convert a resource string to a  sch_resource_t to
- *                      kilobytes
- *                      example: 1mb -> 1024
- *				 	1mw -> 1024 * SIZE_OF_WORD
+ * 		res_to_num - convert a resource string to a sch_resource_t to
+ *                      kilobytes or seconds or 0 for False & 1 for
+ *                      True
```
Comment (bhroam) [inline src/scheduler/misc.c]: So it might return 0 for false and 1 for true, 1 might just be 1 and 0 might just be 0.  Now that I think about this, why don't you just shorten this to 'convert a resource string to a numeric sch_resource_t'

File: src/scheduler/misc.c [hunk 9]
```diff
@@ -242,17 +242,23 @@ add_str_to_array(char ***str_arr, char *str)
...
+ *                      example:
+ *                      1mb -> 1024
+ *				 		1mw -> 1024 * SIZE_OF_WORD
+ *				 		00:00:20 -> 20
+ *				 		True -> 1
+ *				 		False -> 0
```
Comment (bhroam) [inline src/scheduler/misc.c]: The indentation looks off here.  Is it actually off, or is it github?

File: src/scheduler/misc.c [hunk 10]
```diff
@@ -242,17 +242,23 @@ add_str_to_array(char ***str_arr, char *str)
...
  *
  * @param[in]	res_str	-	the resource string
  * @param[out]	type	-	the type of the resource
  *
  * @return	sch_resource_t
  * @retval	a number in kilobytes or seconds
- * @retval	0 for False, 1 for True
+ * @retval	0 for False
+ * @retval	1 for True
```
Comment (bhroam) [inline src/scheduler/misc.c]: For these two say 'if type is boolean'

File: src/scheduler/resource_resv.c [hunk 11]
```diff
@@ -862,7 +862,7 @@ create_resource_req(char *name, char *value)
 			if (value != NULL) {
 				if (set_resource_req(resreq, value) != 1) {
 					schdlog(PBSEVENT_DEBUG, PBS_EVENTCLASS_SCHED, LOG_DEBUG, name,
-						"Bad resource data found");
+						"Bad resource requested data");
```
Comment (bhroam) [inline src/scheduler/resource_resv.c]: Maybe "Bad requested resource data" ?  It sounds funny as it is.

File: test/tests/pbs_smoketest.py [hunk 12]
```diff
@@ -1251,6 +1251,8 @@ def delete_resource_helper(self, r, t, f, c, obj_type, obj_id):
                 if obj_type == JOB:
                     j = Job(TEST_USER1, {'Resource_List.' + r: val})
                 else:
+                    print "======= Resource being set on the reservation:" +\
+                        str(r) + ":" + str(val)
```
Comment (bhroam) [inline test/tests/pbs_smoketest.py]: Don't use print, use self.logger.info() or self.logger.debug()

Comment (agrawalravi90) [reply test/tests/pbs_smoketest.py]: oh my bad, this got left in, I was using it for debugging

**Comment by agrawalravi90** [comment]
Thanks Bhroam, requesting a review from @PBSPro/pbspro-maintainers, thanks!

**Comment by agrawalravi90** [comment]
Requesting one more review please, it's a pretty small code change, thanks!

**Comment by agrawalravi90** [comment]
Requesting one more review please, it's a pretty small code change, thanks! [2]

File: src/scheduler/misc.c [hunk 13]
```diff
@@ -889,43 +888,45 @@ is_num(char *str)
...
 		return -1;
 
 	if (str[0] == '-' || str[0] == '+')
 		str++;
 
-	for (i = 0; i < strlen(str) && (isdigit(str[i]) || str[i] == ':'); i++) {
+	str_len = strlen(str);
+	for (i = 0; i < str_len && (isdigit(str[i]) || str[i] == ':'); i++) {
 		if (str[i] == ':')
 			colen_count++;
```
Comment (mike0042) [inline src/scheduler/misc.c]: It's "colon", not "colen". I know you didn't introduce it, but there should be misspellings like this.

File: src/scheduler/misc.c [hunk 14]
```diff
@@ -889,43 +888,45 @@ is_num(char *str)
...
 	}
 
 	/* is the string completly numeric or a time(HH:MM:SS or MM:SS) */
-	if (i == strlen(str) && colen_count <= 2)
+	if (i == str_len && colen_count <= 2)
```
Comment (mike0042) [inline src/scheduler/misc.c]: It's "colon", not "colen". I know you didn't introduce it, but there should be misspellings like this.

File: src/scheduler/misc.c [hunk 15]
```diff
@@ -889,43 +888,45 @@ is_num(char *str)
...
 		return 1;
 
 	/* is the string a size */
-	if (i == strlen(str) - 2) {
+	if ((i == (str_len - 2)) || (i == (str_len - 1))) {
```
Comment (mike0042) [inline src/scheduler/misc.c]: The comment above this line is pointless. Please rephrase it or remove it.

All this test really tells you is that there is a character that is not a digit or a colon appearing somewhere in the string, not that it is formatted properly for a size. It doesn't distinguish between 123m and 12m3.

Comment (agrawalravi90) [reply src/scheduler/misc.c]: "It doesn't distinguish between 123m and 12m3"
- Actually it does. "12m3" will satisfy the first if (c == 'k' || c == 'm' || c == 'g' || c == 't'). Then the next line c = tolower(str[i+1]); will move the pointer c from 'm' to '3', and then the inner if (c == 'b' || c == 'w' || c == '\0') will not be satisfied for "12m3". It would have ben satisfied if the string was "12mb"/"12mw"/"12m" instead.

The comment above is meant to convey that this check is testing for the string to be a size type resource like mem, I'll clarify that more.

**Comment by agrawalravi90** [comment]
Thanks for approving the changes Mike, I've squashed and rebased, requesting you and @bhroam to merge this in once the CI passes. Thanks!

**Comment by bhroam** [comment]
I merged in the PR.
